Start documentation on confidential (#655)
aleph-im/aleph-vm: doc/confidential.md

# Confidential computing

> Aleph vm offer to launch confidential VM with AMD SEV. This is also know as TEE, Trusted Execution Environment

Reviewer: ```suggestion
Aleph-vm offers to launch confidential VM with AMD SEV. This is also know as TEE, Trusted Execution Environment.
```

> This is only supported for instance with the qemu hypervisor.

Reviewer: ```suggestion
This is only supported for instances with the Qemu as the hypervisor.
```

## Life cycle

> First, a user creates a VM message and send it with notify_allocate. This simply notifies the API of the new VM.  

Reviewer: ```suggestion
First, a user creates a VM message and sends it with notify_allocate. This notifies the orchestrator about the creation of the new VM.  
```

> The user fetch the platform certificate, validate it's chain again AMD root certificate.

Reviewer: ```suggestion
The user fetches the platform certificate, validates it's chain again AMD root certificate.
```

The end result is a virtual machine that is accessible through SSH and is completely encrypted in RAM, making it inaccessible from the point of view of the hypervisor.

```mermaid
flowchart TD
    A[Start] -->|Allocate VM on CRN| B(CRN: Check payment, download image, volume)
    B --> |Download certificate from CRN| C(User: Validate Certificate again CHAIN)
    C --> |Create session certificates| D[Certificates file created]
    D --> |Send certificate to CRN to init sessions | E[CRN: Launch VM with firmware with encrypted communication channel]
    E --> |Fetch measurement from VM| F[User: Calculate it's own measurement and verify them again the CRN's]
    F --> | if ok: Send secret in encrypted channel | G[CRN: Start and unlock VM]
```
    

# CRN Side

## Hardware requirement
An AMD process supporting SEV, minimum Epyc 3

## Requirements for the CRN

> * Support must be enabled in the computer BIOS

Author: not sure the name would vary by BIOS  / motherboard manufacturer and it's Alie who enabled it on our server. But I will try to provide some links

> * The kernel and platform must support SEV.

Author: yes 
This  is actually what we are using for testing.

* sevctl must be installed. A copy is included in the aleph-vm Debian package
* QEMU must be installed.

Check with the `sevctl ok` command that the system is supporting AMD SEV properly, you need at least:

```[ PASS ]   - Secure Encrypted Virtualization (SEV)```

See AMD DOC for more info on enabling SEV for your system
https://www.amd.com/fr/developer/sev.html

## Enabling the confidential computing feature in aleph-vm

Set the following environment variables to launch aleph-vm:
```

> ALEPH_VM_ENABLE_QEMU_SUPPORT=1

Author: it is not at the moment.

 `/etc/systemd/system/aleph-vm-controller@.service` to point  to your version.
 
 For example here is what I use
 ```
 [Unit]
Description=Aleph VM %i Controller Olivier
After=network.target

[Service]
Type=simple
RestartSec=5s
PrivateTmp=yes
NoNewPrivileges=true
WorkingDirectory=/home/olivier/pycharm-aleph-vm/src
Environment=PYTHONPATH=/home/olivier/pycharm-aleph-vm/src:$PYTHONPATH
ExecStart=/home/olivier/.virtualenvs/aleph-vm/bin/python3 -m aleph.vm.controllers --config=/var/lib/aleph/vm/%i-controller.json
Restart=no

[Install]
WantedBy=multi-user.target
```

After modification use the following command to have the modification taken into account
```shell
sudo systemctl daemon-reload
```

# Testing

After initializing the VM you can check it's status with:
`sudo systemctl status aleph-vm-controller@decadecadecadecadecadecadecadecadecadecadecadecadecadecadecadeca.service`

and see the logs with
` sudo journalctl -u aleph-vm-controller@decadecadecadecadecadecadecadecadecadecadecadecadecadecadecadeca.service`

**Important**
 

> If you modify your base image between test you will need to delete the image file on disk (which is a delta of the base image)

Reviewer: ```suggestion
If you modify your base image between tests, you will need to delete the image file on disk (which is a delta of the base image)
```
